flink-formats/flink-avro-glue-schema-registry/src/main/java/org/apache/flink/formats/avro/glue/schema/registry/GlueSchemaRegistryAvroSchemaCoder.java

```diff
@@ -0,0 +1,81 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.avro.glue.schema.registry;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.formats.avro.SchemaCoder;
+
+import org.apache.avro.Schema;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+
+/**
+ * Schema coder that allows reading schema that is somehow embedded into serialized record. Used by
+ * {@link GlueSchemaRegistryAvroDeserializationSchema} and {@link
+ * GlueSchemaRegistryAvroSerializationSchema}.
+ */
+public class GlueSchemaRegistryAvroSchemaCoder implements SchemaCoder {
+    private GlueSchemaRegistryInputStreamDeserializer glueSchemaRegistryInputStreamDeserializer;
+    private GlueSchemaRegistryOutputStreamSerializer glueSchemaRegistryOutputStreamSerializer;
+
+    /**
+     * Constructor accepts transport name and configuration map for AWS Glue Schema Registry.
+     *
+     * @param transportName topic name or stream name etc.
+     * @param configs configurations for AWS Glue Schema Registry
+     */
+    public GlueSchemaRegistryAvroSchemaCoder(
+            final String transportName, final Map<String, Object> configs) {
+        glueSchemaRegistryInputStreamDeserializer =
+                new GlueSchemaRegistryInputStreamDeserializer(configs);
+        glueSchemaRegistryOutputStreamSerializer =
+                new GlueSchemaRegistryOutputStreamSerializer(transportName, configs);
+    }
+
+    @VisibleForTesting
+    protected GlueSchemaRegistryAvroSchemaCoder(
+            final GlueSchemaRegistryInputStreamDeserializer
+                    glueSchemaRegistryInputStreamDeserializer) {
+        this.glueSchemaRegistryInputStreamDeserializer = glueSchemaRegistryInputStreamDeserializer;
+    }
+
+    @VisibleForTesting
+    protected GlueSchemaRegistryAvroSchemaCoder(
+            final GlueSchemaRegistryOutputStreamSerializer
+                    glueSchemaRegistryOutputStreamSerializer) {
+        this.glueSchemaRegistryOutputStreamSerializer = glueSchemaRegistryOutputStreamSerializer;
+    }
+
+    @Override
+    public Schema readSchema(InputStream in) throws IOException {
+        return glueSchemaRegistryInputStreamDeserializer.getSchemaAndDeserializedStream(in);
+    }
+
+    @Override
+    public void writeSchema(Schema schema, OutputStream out) throws IOException {
+        byte[] data = ((ByteArrayOutputStream) out).toByteArray();
```

Comment:
It's always `ByteArrayOutputStream` since the stream is from `GlueSchemaRegistrySerializationSchema` and it's defined as `ByteArrayOutputStream` in super class  `AvroSerializationSchema`.

Reply:
ok, please add the `Preconditions` check

Reply:
The Precondition check should go before the cast

Reply:
@LinyuYao1021 this comment is still open